api/src/test/java/org/openmrs/api/TagServiceTest.java

```diff
@@ -0,0 +1,145 @@
+import org.openmrs.api.context.Context;
+import org.openmrs.api.db.TagDAO;
+import org.openmrs.api.impl.TagServiceImpl;
+import org.openmrs.test.BaseModuleContextSensitiveTest;
+import sun.reflect.annotation.ExceptionProxy;
+
+import javax.validation.constraints.AssertTrue;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.mockito.Mockito.*;
+import static org.hamcrest.Matchers.*;
+import static org.junit.Assert.*;
+
+/**
+ * This is a unit test, which verifies logic in TagService.
+ */
+public class TagServiceTest extends BaseModuleContextSensitiveTest {
+	
+	private static final String TAG_INITIAL_XML = "TagServiceDataset.xml";
+	
+	@InjectMocks
+	TagServiceImpl basicModuleService;
+	
+	@Mock
+	TagDAO dao;
+	
+	@Mock
+	UserService userService;
+	
+	@Mock
+	ConceptService conceptService;
+	
+	@Before
+	public void setupMocks() {
+		MockitoAnnotations.initMocks(this);
+	}
+	
+	@Test
+	public void getAlltags_shouldFetchAllTags() throws Exception {
+		executeDataSet(TAG_INITIAL_XML);
+		List<String> list = Context.getService(TagService.class).getAllTags();
+		assertEquals(13, list.size());
+	}
+	
+	@Test
+	public void getTag_shouldFetchUniqueMatchingTag() throws Exception {
+		executeDataSet(TAG_INITIAL_XML);
+		Tag tag = Context.getService(TagService.class).getTag(3);
+		assertEquals(tag.getUuid(), "e12c432c-1b9f-343e-b332-f3ef6c88ad3f");
+	}
+	
+	@Test
+	public void getTags_shouldFetchListOfMatchingTags() throws Exception {
+		executeDataSet(TAG_INITIAL_XML);
+		List<Tag> tagList = Context.getService(TagService.class).getTags("FollowUp");
+		assertEquals(tagList.size(), 4);
+	}
+	
+	@Test
+	public void object_existsMethod_shouldReturnTrueIfObjectExists() throws Exception {
+		OpenmrsObject object = Context.getService(TagService.class).getObject(Encounter.class,
+		    "e403fafb-e5e4-42d0-9d11-4f52e89d148c");
+		Encounter encounter = Context.getEncounterService().getEncounterByUuid("e403fafb-e5e4-42d0-9d11-4f52e89d148c");
+		assertEquals(object, encounter);
+	}
+	
+	@Test
+	public void getTagsWithObjectParameter_shouldReturnListOfStringTags() throws Exception {
+		executeDataSet(TAG_INITIAL_XML);
+		Encounter encounter = Context.getEncounterService().getEncounterByUuid("e403fafb-e5e4-42d0-9d11-4f52e89d148c");
+		List<Tag> tags = Context.getService(TagService.class).getTags(encounter);
+		assertEquals(tags.size(), 3);
+	}
+	
+	@Test
+	public void getObjectsWithAllTags_shouldReturnOpenmrsObjectsWhichHaveAllTags() throws Exception {
+		executeDataSet(TAG_INITIAL_XML);
+		List<String> types = new ArrayList<String>();
+		types.add("org.openmrs.Encounter");
+		types.add("org.openmrs.Obs");
+		List<String> tags = new ArrayList<String>();
+		tags.add("Initial");
+		tags.add("FollowUp");
+		List<OpenmrsObject> tagList = Context.getService(TagService.class).getObjectsWithAllTags(types, tags);
+		assertEquals(tagList.size(), 3);
+	}
+	
+	@Test
+	public void getTags_shouldReturnTagsWhichMatchParameterCriteria() throws Exception {
```

Comment:
this tests the getTags method, which return tag objects of those which have objectType either Obs or Order, and have tag either Initial or FollowUp. there are 2 such Tag objects for 1 Obs object and 1 tag object for 1 order object